Add more visibility to admins (part-2)

rt4914 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://developers.forem.com/contributing-guide/forem#create-a-pull-request
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [x] Feature

## Description

Add a section to the Article Admin page found here `/admin/content_manager/articles/{article-id}`.

**Overview of Privileged actions with article title**:
* No. of Privileged actions at the top right
* Display the same overview in `/admin/content_manager/articles` as well
* **Score calculation**
  * The top line score represented by the article’s `article.score`
  * More granular visibility into the main inputs that go into that score…
  * The sum of reactions _on the author_, basically, taking the calculation from [this file](https://github.com/forem/forem/blob/main/app/models/article.rb#L559)
  * An aggregate sum point total of all “regular” reactions (❤️, 🦄, 🤯, 🙌, 🔥)
  * An aggregate sum point total of all “privileged” actions (👍, 👎, and 🧐)

**Privileged Actions log**:
* A list of all 👍, 👎, and 🧐 reactions (privileged actions) - We will divide them into two tabs called `Flags` and `Quality reactions`. Note that 🧐 is called `vomit`
* **Flag tab**
  * Column representing the reaction category – reaction.category 🧐
  * Column representing the user who created that reaction – reactions.user_id
  * Column representing the point total of that individual reaction – reactions.points
  * Columns representing the `status` of each privileged reaction. Will be `valid`, `invalid`, `Open`
    * Open -> Admin hasn't marked that action as valid/invalid
    * Valid -> Admin marked that action as valid
    * Invalid -> Admin marked that action as invalid
  * The ability to change the `status` of each action by the three-dot menu on the right.



* **Quality Reactions tab**
  * Column representing the reaction category – reaction.category 👍 👎
  * Column representing the user who created that reaction – reactions.user_id
  * Column representing the point total of that individual reaction – reactions.points
  * Quality reactions will be the log of reactions only, and the admin won't be able to adjust them.

   
**For more details and mocks,** [**check Figma**](https://www.figma.com/file/yhg9HTZDFIeWuvBt9ygrss/Moderation?node-id=1485-4141&t=Pn3Q2r8QQoGd5Vqj-4).

## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Closes #19314

## QA Instructions, Screenshots, Recordings


[**check Figma**](https://www.figma.com/file/yhg9HTZDFIeWuvBt9ygrss/Moderation?node-id=1485-4141&t=Pn3Q2r8QQoGd5Vqj-4).

All the details related to what this PR is, how to test it and what to expect has been covered in this video: 
[**How to test and what to expect?**](https://www.loom.com/share/4813e45d54104322a2f8bd7fa8b08624)


<img width="889" alt="Screenshot 2023-05-04 at 5 57 00 PM" src="https://github.com/forem/forem/assets/9396084/42d0d918-f531-4aa1-b671-84b8e7ebbc6f">


### Output
<img width="500" alt="Screenshot 2023-05-02 at 5 33 19 PM" src="https://user-images.githubusercontent.com/9396084/235661242-e099693c-cde7-4245-b22a-466890c175d6.png">
<img width="500" alt="Screenshot 2023-05-02 at 5 33 28 PM" src="https://user-images.githubusercontent.com/9396084/235661274-3ae7d60d-6e1a-4507-87dd-b725c0cc80f6.png">
<img width="500" alt="Screenshot 2023-05-02 at 5 33 34 PM" src="https://user-images.githubusercontent.com/9396084/235661279-2bf37d05-a6e5-40d4-8f99-befc6fba8b7a.png">
<img width="500" alt="Screenshot 2023-05-02 at 5 33 48 PM" src="https://user-images.githubusercontent.com/9396084/235661283-7cd40fc7-74d7-435a-84e8-1f83dbf77e83.png">
<img width="500" alt="Screenshot 2023-05-02 at 5 34 11 PM" src="https://user-images.githubusercontent.com/9396084/235661289-35dfb83c-c47c-4190-8029-1253132bc744.png">
<img width="500" alt="Screenshot 2023-05-02 at 5 34 19 PM" src="https://user-images.githubusercontent.com/9396084/235661298-bbd60bf3-9837-42f6-9869-8b33bfe1ca54.png">
<img width="500" alt="Screenshot 2023-05-04 at 3 07 53 PM" src="https://user-images.githubusercontent.com/9396084/236168615-73722762-da54-45aa-9af9-4f044ded4e40.png">




<img width="961" alt="Screenshot 2023-05-02 at 5 42 03 PM" src="https://user-images.githubusercontent.com/9396084/235664116-8d31c306-cb24-473b-a150-f0277519af2c.png">
<img width="796" alt="Screenshot 2023-05-02 at 5 45 35 PM" src="https://user-images.githubusercontent.com/9396084/235664130-5af71f3b-0982-479d-90b0-bc4c63258070.png">
<img width="787" alt="Screenshot 2023-05-02 at 5 45 41 PM" src="https://user-images.githubusercontent.com/9396084/235664134-c2f5b8b6-d1a0-4ee8-a5dd-591a837bafdc.png">
<img width="810" alt="Screenshot 2023-05-02 at 5 48 27 PM" src="https://user-images.githubusercontent.com/9396084/235664150-dc905609-76de-4f54-9910-bb2714893620.png">

### UI accessibility concerns?

No

## Added/updated tests?

- [x] No

## [optional] Are there any post deployment tasks we need to perform?

@maestromac We should deploy this on canary1 or canary2 so that admin team can check the feature and provide suggestions.

rt4914 commented:
In this I am facing following error but unable to fix it correct.

```
GET /admin/content_manager/articles/1
| USE eager loading detected
|   Reaction => [:user]
|   Add to your query: .includes([:user])
| Call stack
|   …/forem/app/views/admin/articles/_flag_action_item.html
```

I tried solving this by adding following line:
`Bullet.add_safelist(type: :unused_eager_loading, class_name: "Reaction", association: :user)`
(Similar to https://github.com/forem/forem/pull/19274/files)

But this did not fix the issue.

rt4914 commented:
I have created this detailed video to show exactly what are the problems in my current implementation and what all things I have tried

https://www.loom.com/share/5a9f4e22012b458081b132ffd54e79fd

rt4914 commented:
Converted this PR to draft because there are some logical issues which needs to be fixed first.

rt4914 commented:
@maestromac Please deploy this on Canary1 or canary2 so that we can start testing it. Thanks.

Ridhwana commented on app/views/admin/articles/_article_item.html.erb:
```diff
@@ -0,0 +1,191 @@
+          <span class="c-indicator c-indicator--danger">Unpublished</span>
+        <% end %>
+        <% if decorated_article.pinned? %>
+          <span class="c-indicator c-indicator--warning" data-testid="pinned-indicator">Pinned</span>
```
Do you think we should add these and some others to the i81n file?

rt4914 replied:
Applied this change. Shifted strings to yml files.

rt4914 commented:
@anujbhavsar96 I have removed the time as it was incorrect and also fixed the width or user name
<img width="1001" alt="Screenshot 2023-05-10 at 11 59 44 AM" src="https://github.com/forem/forem/assets/9396084/ce237567-bc05-4beb-be33-cbf026c97e91">

anujbhavsar96 commented:
@rt4914, it would be great if we could display the date there. Of course, time is not that important, but the date can be helpful for admins to judge the overall impression of the article based on recent actions.

rt4914 commented:
> @rt4914, it would be great if we could display the date there. Of course, time is not that important, but the date can be helpful for admins to judge the overall impression of the article based on recent actions.

@anujbhavsar96 I have restored the date and time and made sure that when the status gets updated, the item will show the updated time too.
<img width="972" alt="Screenshot 2023-05-10 at 2 14 30 PM" src="https://github.com/forem/forem/assets/9396084/75b2caf1-34dd-4e25-a754-bab0773901f9">

Ridhwana commented:
@rt4914 just wanted to check if this is ready for review?

Ridhwana commented:
> @Ridhwana From implementation perspective its finished. I am currently working on writing cypress tests for this.

Cool, could we move it into draft until those specs are done? 

> Also, I do have a query, I have pushed my code which contains a `qualityReactionArticle.spec.js`, basically as soon as I create a new article I want that there are some marked privileged reactions so that the vomits, high-quality and low-quality data is pre-filled. Any idea how can I do that? some sort of pre-created database?

If you need to add any seeds, you can do so in [the seeds_e2e](https://github.com/forem/forem/blob/b4a4b30153534ee7c9462d8417d74b4e33bf9e77/spec/support/seeds/seeds_e2e.rb) for Cypress/

maestromac commented on cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js:
```diff
@@ -0,0 +1,181 @@
+    cy.get('.flex .crayons-card:nth-child(2) .fs-s').should('contain', '-50');
+  });
+
+  it('should open the dropdown on button click', () => {
+    cy.get('.c-btn--icon-alone').click();
+    cy.get('.crayons-dropdown').should('be.visible');
+  });
+
+  it('should open the dropdown on button click and close the dropdown on same button click', () => {
+    cy.get('.c-btn--icon-alone').click();
+    cy.get('.crayons-dropdown').should('be.visible');
+
+    cy.get('.c-btn--icon-alone').click();
+    cy.get('.crayons-dropdown').should('not.visible');
+  });
```
E2E is way slower than unit test, so it's more acceptable to test more than one aspect per block. In this case, your `should open the dropdown on button click and close the dropdown on same button click` is already also testing what `should open the dropdown on button click` is doing, making the line 125 redundant.

rt4914 replied:
Updated the test to write it like this:
```
  it('should close the dropdown on same button click twice', () => {
    cy.get('.c-btn--icon-alone').click();
    cy.get('.c-btn--icon-alone').click();
    cy.get('.crayons-dropdown').should('not.visible');
  });
```

maestromac commented on cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js:
```diff
@@ -0,0 +1,181 @@
+    cy.get('.crayons-dropdown').should('not.visible');
+  });
+
+  it('should open the dropdown on button click', () => {
+    cy.get('.c-btn--icon-alone').click();
+    cy.get('.crayons-dropdown').should('be.visible');
+  });
+
+  it('should open the dropdown on button click and close the dropdown on same button click', () => {
+    cy.get('.c-btn--icon-alone').click();
+    cy.get('.crayons-dropdown').should('be.visible');
+
+    cy.get('.c-btn--icon-alone').click();
+    cy.get('.crayons-dropdown').should('not.visible');
+  });
+
```
Are these test duplicate of the above?

rt4914 replied:
They were partially duplicate which I have update now. But the main point is that in first test we are performing single click on button to open the dropdown and in second we are clicking it twice to make sure that the dropdown closes too.

maestromac commented on cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js:
```diff
@@ -0,0 +1,181 @@
+    });
+  });
+
+  it('should display the correct default data', () => {
+    cy.get('.flex .crayons-card:nth-child(3) .fs-s').should('contain', '1');
+  });
+
+  it('should display the correct user name', () => {
+    cy.get('.flex .crayons-subtitle-3').should(
+      'contain',
+      'Trusted User 1 \\:/',
+    );
+  });
+
+  it('should display the open flag status', () => {
+    cy.get('.flex .c-indicator').should('be.visible').as('flagStatus');
+    cy.get('@flagStatus')
+      .should('have.class', 'c-indicator--warning')
+      .and('contain', 'Open');
+  });
+
+  it('should display the correct score for open flag', () => {
+    cy.get('.flex .crayons-card:nth-child(2) .fs-s').should('contain', '-50');
+  });
```
E2E is way slower than unit test, so it's more acceptable to test more than one aspect per block. In this case, these 4 specs can be merged into one and you can use comments to say what you are testing. ie
```
  it('should display the correct default data', () => {
    // show corrects user name
    ...

     // shows correct flag
     ...
  })
```

This way you check for all things in a single page load.

rt4914 replied:
Thanks for this suggestion. Based on this I have merged these tests to reduce them to almost half.

anujbhavsar96 reviewed (approved):
Everything works as expected 🚀. Great work @rt4914 ✨

rt4914 commented:
@Ridhwana @filleduchaos Looking for review on this PR. Thanks :-)

Ridhwana commented:
@rt4914 there were some comments from Mac on the PR, are you still addressing those or can I go ahead and review?

rt4914 commented:
> @rt4914 there were some comments from Mac on the PR, are you still addressing those or can I go ahead and review?

@Ridhwana I completely missed those 3 comments in collapsed state. Just addressing them, overall they are not a blocker as such.

Ridhwana reviewed (approved):
LGTM 🚀

maestromac reviewed (approved):
Thanks for updating the test. LGTM!
